pkg/controllers/provisioning/v1alpha1/reallocator/filter.go

```diff
@@ -25,40 +28,75 @@ type Filter struct {
 	kubeClient client.Client
 }
 
-// Gets Nodes fitting some underutilization
-// TODO: add predicate
-func (f *Filter) GetUnderutilizedNodes(ctx context.Context) ([]*v1.Node, error) {
-	nodeList := &v1.NodeList{}
+// Gets Nodes fitting some underutilization predicate
+func (f *Filter) GetUnderutilizedNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
 	underutilized := []*v1.Node{}
 
-	// 1. Get all nodes
-	if err := f.kubeClient.List(ctx, nodeList); err != nil {
-		return nil, fmt.Errorf("listing nodes, %w", err)
+	// 1. Get all provisioner labeled nodes
+	nodeList, err := f.getNodes(ctx, provisioner)
+	if err != nil {
+		return nil, err
 	}
 
 	// 2. Get nodes and the pods on each node
 	for _, node := range nodeList.Items {
 		pods, err := f.getPodsOnNode(ctx, node.Name)
 		if err != nil {
-			return []*v1.Node{}, fmt.Errorf("filtering pods on nodes, %w", err)
+			return []*v1.Node{}, err
 		}
-		if f.isUnderutilized(&node, pods) {
+		// Only checks if it has 0 non-daemon pods right now
+		if f.isUnderutilized(pods) {
 			underutilized = append(underutilized, &node)
 		}
 	}
 	return underutilized, nil
 }
 
+func (f *Filter) GetExpiredNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) ([]*v1.Node, error) {
+	nodeList, err := f.getNodes(ctx, provisioner)
+	if err != nil {
+		return nil, err
+	}
+	nodes := []*v1.Node{}
+	for _, node := range nodeList.Items {
+		ok, err := nodeUtil.IsPastTTL(&node)
+		if err != nil {
+			return nodes, fmt.Errorf("checking TTL, %w", err)
+		}
+		if ok {
+			nodes = append(nodes, &node)
+		}
+	}
+	return nodes, nil
+}
+
+func (f *Filter) getNodes(ctx context.Context, provisioner *v1alpha1.Provisioner) (*v1.NodeList, error) {
+	nodeList := &v1.NodeList{}
+	if err := f.kubeClient.List(ctx, nodeList, client.MatchingLabels(map[string]string{
+		v1alpha1.ProvisionerNameLabelKey:      provisioner.Name,
+		v1alpha1.ProvisionerNamespaceLabelKey: provisioner.Namespace,
+	})); err != nil {
+		return nil, fmt.Errorf("listing nodes, %w", err)
+	}
+	return nodeList, nil
```

Comment:
Would it simplify things to return nodeList.Items here?

Reply:
I see what you're saying, and I had that thought too. `nodeList.Items` is a `[]Node` not a `[]*Node` which would require some refactoring on all of the functions. I think it's fine like this.

Reply:
[]Node is actually quite performant, since it's a pointer. Don't let the existing (tiny) codebase bully your implementation decisions to be worse.